### PR TITLE
perf: dont rerender all modals when opening/closing one

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Improves UI performance with 2 main changes:
- Splits up state management of modals such that they dont all rerender when one of them opens or closes
- Creates separate instances of AssetSearch for buy and sell asset selection so interacting with it doesnt trigger rerendering (reuse of the modal changes props which results in rerenders)

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

NA

## Risk

Moderate risk of things breaking for the modals across the app, especially keepkey wiping which seems fragile judging by the comments in the codebase.

## Testing

See above

### Engineering

See above

### Operations

See above

### CODE REVIEW

**Comment by woodenfurniture** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4969** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4969" target="_blank">Graphite</a> 
    * **PR #4971** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4971" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/4971?utm_source=stack-comment).

File: src/components/Modals/AssetSearch/AssetSearchModal.tsx [hunk 1]
```diff
@@ -17,23 +17,28 @@ import { useWindowSize } from 'hooks/useWindowSize/useWindowSize'
 import type { Asset } from 'lib/asset-service'
 import { breakpoints } from 'theme/theme'
 
-interface AssetSearchModalProps extends AssetSearchProps {
-  onClick: Required<AssetSearchProps>['onClick']
+export type AssetSearchModalProps = AssetSearchProps & {
   title?: string
+  onClick: Required<AssetSearchProps>['onClick']
```
Comment (0xdef1cafe) [inline src/components/Modals/AssetSearch/AssetSearchModal.tsx:22]: TIL

File: src/components/Modals/FeedbackSupport/FeedbackSupport.tsx [hunk 2]
```diff
@@ -58,7 +56,6 @@ export const FeedbackAndSupport = () => {
               as={Link}
               size='sm'
               label={translate('common.submitFeatureRequest')}
-              isExternal
```
Comment (gomesalexandre) [inline src/components/Modals/FeedbackSupport/FeedbackSupport.tsx:61]: Keep one of these two?

**Review by 0xdef1cafe — APPROVED** [review]
gave this a gomesque review

tested all modals except keepkey wipe as i can't find one currently

typing works as expected and it's a 3 line net negative diff

File: src/context/ModalProvider/types.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import type { SendModalProps } from 'components/Modals/Send/Send'
+
+import type { CLOSE_MODAL, OPEN_MODAL } from './constants'
+
+export type Modals = {
+  receive: ({ asset, accountId }: ReceivePropsType) => JSX.Element
+  qrCode: ({ assetId, accountId }: QrCodeModalProps) => JSX.Element
+  send: ({ assetId, accountId, input }: SendModalProps) => JSX.Element
+  fiatRamps: FC<FiatRampsModalProps>
+  settings: () => JSX.Element
```
Comment (gomesalexandre) [inline src/context/ModalProvider/types.ts:18]: All modals are functional components and could be typed as such e.g
```suggestion
  settings: FC
```

File: src/context/ModalProvider/types.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  keepKeyWipe: () => JSX.Element
+  backupNativePassphrase: FC<BackupPassphraseModalProps>
+  mobileWelcomeModal: () => JSX.Element
+  addAccount: () => JSX.Element | null
```
Comment (gomesalexandre) [inline src/context/ModalProvider/types.ts:22]: Ditto for this one, `React.FC` infers this if we want to use it instead for consistency
image

File: src/context/ModalProvider/ModalContext.tsx [hunk 5]
```diff
@@ -1,10 +1,13 @@
...
-// If initial state is removed/set to null, the KeepKey wallet modals will break
-export const ModalContext = createModalContext<ModalStateType>()
+import { MODAL_KEYS } from './constants'
+import type { BaseProps, ModalContext, Modals } from './types'
+
+export const modalContext: ModalContext = MODAL_KEYS.reduce<ModalContext>(
+  (acc, key: keyof Modals) => ({
+    ...acc,
+    // blatant cast as we inject the correct props during provider creation in ModalContainer
+    [key]: createContext<BaseProps<keyof Modals>>({} as BaseProps<keyof Modals>),
```
Comment (gomesalexandre) [inline src/context/ModalProvider/ModalContext.tsx:10]: 🧠

**Review by gomesalexandre — APPROVED** [review]
Confirmed modals aren't breaking locally and are snappy on my turbo slow dev server. 
Also confirmed you're a certified gangster ser

**Pull request merged**